### ISSUE TITLE
Ambiguous occurrence ‘defaultTimeLocale’

### ISSUE THREAD

**Issue by ronin13** [jaor/xmobar#213]
[ 1 of 51] Compiling StatFS           ( dist/build/xmobar/xmobar-tmp/StatFS.hs, dist/build/xmobar/xmobar-tmp/StatFS.o ) [Data.ByteString.Char8 changed]
    [ 2 of 51] Compiling Actions          ( src/Actions.hs, dist/build/xmobar/xmobar-tmp/Actions.o ) [System.Process changed]
    [ 3 of 51] Compiling MinXft           ( dist/build/xmobar/xmobar-tmp/MinXft.hs, dist/build/xmobar/xmobar-tmp/MinXft.o ) [Graphics.X11 changed]
    [ 4 of 51] Compiling ColorCache       ( src/ColorCache.hs, dist/build/xmobar/xmobar-tmp/ColorCache.o ) [Graphics.X11.Xlib changed]
    [ 6 of 51] Compiling Signal           ( src/Signal.hs, dist/build/xmobar/xmobar-tmp/Signal.o ) [dist/build/autogen/cabal_macros.h changed]
    [ 7 of 51] Compiling XUtil            ( dist/build/xmobar/xmobar-tmp/XUtil.hs, dist/build/xmobar/xmobar-tmp/XUtil.o ) [Graphics.X11.Xlib changed]
    [ 8 of 51] Compiling Commands         ( src/Commands.hs, dist/build/xmobar/xmobar-tmp/Commands.o ) [XUtil changed]
    [ 9 of 51] Compiling Plugins          ( src/Plugins.hs, dist/build/xmobar/xmobar-tmp/Plugins.o ) [Commands changed]
    [10 of 51] Compiling Plugins.CommandReader ( src/Plugins/CommandReader.hs, dist/build/xmobar/xmobar-tmp/Plugins/CommandReader.o )
    [11 of 51] Compiling Paths_xmobar     ( dist/build/autogen/Paths_xmobar.hs, dist/build/xmobar/xmobar-tmp/Paths_xmobar.o )
    [12 of 51] Compiling Plugins.Date     ( src/Plugins/Date.hs, dist/build/xmobar/xmobar-tmp/Plugins/Date.o )

    src/Plugins/Date.hs:36:33:
        Ambiguous occurrence ‘defaultTimeLocale’
        It could refer to either ‘System.Locale.defaultTimeLocale’,
                                imported from ‘System.Locale’ at src/Plugins/Date.hs:23:1-20
                            or ‘Data.Time.defaultTimeLocale’,
                                imported from ‘Data.Time’ at src/Plugins/Date.hs:25:1-16
                                (and originally defined in ‘time-1.5.0.1:Data.Time.Format.Locale’)
    Failed to install xmobar-0.23
    cabal: Error: some packages failed to install:
    xmobar-0.23 failed during the building phase. The exception was:
    ExitFailure 1
    cabal install --upgrade-dependencies --flags='with_xft with_utf8 with_alsa'  117.06s user 3.44s system 98% cpu 2:02.86 total

Full log here:  http://paste.wnohang.net/8d0996

Fixed with:

    diff --git a/src/Plugins/Date.hs b/src/Plugins/Date.hs
    index a263536..ac8efda 100644
    --- a/src/Plugins/Date.hs
    +++ b/src/Plugins/Date.hs
    @@ -20,7 +20,7 @@ module Plugins.Date (Date(..)) where

    import Plugins

    -import System.Locale
    +import System.Locale hiding (defaultTimeLocale)
    import Control.Monad (liftM)
    import Data.Time

Versions:

    ghc -V
    The Glorious Glasgow Haskell Compilation System, version 7.8.4

    cabal --version
    cabal-install version 1.22.0.0
    using version 1.22.0.0 of the Cabal library

**Comment by ronin13** [comment]
Better fix:

    diff --git a/src/Plugins/Date.hs b/src/Plugins/Date.hs
    index a263536..37e603f 100644
    --- a/src/Plugins/Date.hs
    +++ b/src/Plugins/Date.hs
    @@ -20,7 +20,6 @@ module Plugins.Date (Date(..)) where

    import Plugins

    -import System.Locale
    import Control.Monad (liftM)
    import Data.Time

**Comment by jaor** [comment]
@ronin13 thanks for the heads up.  AFAICT, the fix is only applicable for 7.8.4, which i cannot test right now, but i'll add your fix with a CPP conditional.

**Comment by jaor** [comment]
@ronin13 i've just added your fix for GHC > 7.8.4.  please let me know if it works for you, thanks!

**Comment by obszczymucha** [comment]
I'm not sure if this is related, but I just downloaded the latest version from github and tried to compile it and got:
```
cabal install --upgrade-dependencies --flags="with_xft with_alsa with_dbus"
Resolving dependencies...
Configuring xmobar-0.23...
Building xmobar-0.23...
Failed to install xmobar-0.23
Build log ( /home/alien/.cabal/logs/xmobar-0.23.log ):
Configuring xmobar-0.23...
Building xmobar-0.23...
Preprocessing executable 'xmobar' for xmobar-0.23...
[12 of 52] Compiling Plugins.Date     ( src/Plugins/Date.hs, dist/build/xmobar/xmobar-tmp/Plugins/Date.o )

src/Plugins/Date.hs:41:33: Not in scope: ‘defaultTimeLocale’
cabal: Error: some packages failed to install:
xmobar-0.23 failed during the building phase. The exception was:
ExitFailure 1
```

``src/Plugins/Date.hs`` contained this:
```
#ifdef GHC783
import System.Locale
#endif
```

When I commented out the ``#ifdef`` directive, it compiled properly. I'm on GHC 7.8.4.

**Comment by jaor** [comment]
sorry, that's unrelated.  i was trying to fix an unrelated issue after
the changes for the alsa one.  i've reverted the offending code.

**Comment by jaor** [comment]
@obszczymucha er, i mean, that's related (i thought i was answering to a different bug! :)) so even in 7.8.4 we need `defaultTimeLocale`, so i'm not sure where @ronin13's error comes from...

**Comment by ronin13** [comment]
I think the error stemmed from this -     "Ambiguous occurrence ‘defaultTimeLocale’" since time-1.5.0.1 also provides that now. The time package in ghc-7.8.4 doesn't have this. (on Arch)

```
ghc-pkg list time
/usr/lib/ghc-7.8.4/package.conf.d
   time-1.4.2
/home/raghavendra/.ghc/x86_64-linux-7.8.4/package.conf.d
   time-1.5.0.1
   
```

**Comment by jaor** [comment]
i see... it seems we cannot make xmobar depend on time-1.5 because we
depend on timezone-series 0.1.*, which requires time < 1.5.  i guess
depending explicitly on time == 1.4.* would still break compilation in
ghc 7.8.4?

**Comment by ronin13** [comment]
I suppose that may work since I couldn't find anything that specifically 
requires time>=1.5.*

However, I did see this, but this is due to how cabal deps work, many of them 
just have time>1.4 or so. 

```
ghc-pkg unregister --user time-1.5.0.1
ghc-pkg: unregistering time-1.5.0.1 would break the following packages: Cabal-1.22.1.1 optparse-applicative-0.11.0.2 ansi-wl-pprint-0.6.7.1 tar-0.4.1.0 ansi-terminal-0.6.2.1 xmonad-contrib-0.12 random-1.1 xmonad-0.12 HTTP-4000.2.19 network-2.6.0.2 alsa-mixer-0.2.0.2 process-1.2.1.0 directory-1.2.1.0 unix-2.7.1.0 (use --force to override)
```


I am not sure how cabal dependencies work with actual builds, will time == 1.4.* in cabal hide time-1.5.0.1?  (here hide/expose in ghc-pkg's parlance). (or do you mean enforcing dependency in code).


cabal/ghc allows for multiple time packages at same time (as in previous 
comment), so xmobar should be 
fine.

**Comment by jaor** [comment]
i meant time == 1.4.* in the xmobar.cabal file... not sure either how
that's handled by cabal...

**Comment by ronin13** [comment]
time-1.5.0.1  is installed on a few distros as per http://hackage.haskell.org/package/time (NixOS, Fedora etc.), so you can check how they are managing it with xmobar.

**Comment by ronin13** [comment]
This breaks builds with ghc 7.10 since it ships with time-1.5.0.1.

**Comment by jaor** [comment]
does PR #224, which i just merged, fix this issue for you?

**Comment by ronin13** [comment]
Yes, it does. Thanks.
